### PR TITLE
HW6 3.3 Implemented flower store

### PR DESCRIPTION
**Pull request by taras309**

### CODE REVIEW

File: src/flowerstore/Chamomile.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package flowerstore;
```
Comment (Anderwerty) [inline src/flowerstore/Chamomile.java]: Let's make root package not flowerstore, but com .store.flowers.
And put  all flower class into package  com .store.flowers.entity or  com .store.flowers.domain

File: src/flowerstore/Flower.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package flowerstore;
+
+public abstract class Flower {
```
Comment (Anderwerty) [inline src/flowerstore/Flower.java]: Please add field price into this class and constructor with the parameter.

File: src/flowerstore/FlowerStore.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class FlowerStore {
+
+    private int wallet;
+
+    public List<Flower> sell(int numRoses, int numChamomiles, int numTulips) {
```
Comment (Anderwerty) [inline src/flowerstore/FlowerStore.java]: According to the task, this method should return array not list

File: src/flowerstore/FlowerStore.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        guardFromNegative(numRoses, numChamomiles, numTulips);
+
+        return Stream.of(flowerStreamOf(Rose::new, numRoses), flowerStreamOf(Chamomile::new, numChamomiles),
+                flowerStreamOf(Tulip::new, numTulips))
+                .flatMap(flower -> flower)
+                .peek(this::addPriceToWallet)
+                .collect(Collectors.toList());
+    }
+
+    public List<Flower> sellSequence(int numRoses, int numChamomiles, int numTulips) {
```
Comment (Anderwerty) [inline src/flowerstore/FlowerStore.java]: the same

File: src/flowerstore/FlowerStore.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
+package flowerstore;
```
Comment (Anderwerty) [inline src/flowerstore/FlowerStore.java]: package flowerstore; -> package com.store.flowers.service

File: src/flowerstore/IllegalNumberOfFlowersException.java [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package flowerstore;
```
Comment (Anderwerty) [inline src/flowerstore/IllegalNumberOfFlowersException.java]: package flowerstore; -> package com.store.flowers.exception

File: src/flowerstore/Main.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package flowerstore;
```
Comment (Anderwerty) [inline src/flowerstore/Main.java]: package flowerstore; -> package com.store.flowers;

File: src/flowerstore/Main.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+
+import java.util.List;
+
+public class Main {
```
Comment (Anderwerty) [inline src/flowerstore/Main.java]: Main -> ConsoleApplication

File: src/com/store/flowers/Chamomile.java [hunk 9]
```diff
@@ -0,0 +1,15 @@
+package com.store.flowers;
```
Comment (Anderwerty) [inline src/com/store/flowers/Chamomile.java]: package com.store.flowers; -> package com.store.flowers.entity;

File: src/com/store/flowers/Flower.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package com.store.flowers;
```
Comment (Anderwerty) [inline src/com/store/flowers/Flower.java]: package com.store.flowers; -> package com.store.flowers.entity;

File: src/com/store/flowers/FlowerStore.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
+package com.store.flowers;
```
Comment (Anderwerty) [inline src/com/store/flowers/FlowerStore.java]: package com.store.flowers;  -> package com.store.flowers.service;

File: src/com/store/flowers/FlowerStore.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+public class FlowerStore {
+
+    private int wallet;
+
+    public Flower[] sell(int numRoses, int numChamomiles, int numTulips) {
+        guardFromNegative(numRoses, numChamomiles, numTulips);
+
+        return Stream.of(flowerStreamOf(this::getRose, numRoses), flowerStreamOf(this::getChamomile, numChamomiles),
+                flowerStreamOf(this::getTulip, numTulips))
+                .flatMap(flower -> flower)
```
Comment (Anderwerty) [inline src/com/store/flowers/FlowerStore.java]: Do you need this method?

File: src/com/store/flowers/FlowerStore.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+                Flower tulip = tulips.next();
+                addPriceToWallet(tulip);
+                flowers[++index] = tulip;
+            }
+        }
+
+        return flowers;
+    }
+
+    public int getWallet() {
```
Comment (Anderwerty) [inline src/com/store/flowers/FlowerStore.java]: Move it and put after field.` private int wallet;`

File: src/com/store/flowers/Tulip.java [hunk 14]
```diff
@@ -0,0 +1,15 @@
+package com.store.flowers;
```
Comment (Anderwerty) [inline src/com/store/flowers/Tulip.java]: package com.store.flowers; -> package com.store.flowers.entity;

File: src/com/store/flowers/Rose.java [hunk 15]
```diff
@@ -0,0 +1,15 @@
+package com.store.flowers;
```
Comment (Anderwerty) [inline src/com/store/flowers/Rose.java]: The same

File: src/com/store/flowers/Chamomile.java [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+
+public class Chamomile extends Flower {
+
+    private int price;
```
Comment (Anderwerty) [inline src/com/store/flowers/Chamomile.java]: It looks like a copy/past. You put the same field in all classes of this hierarchy. So, put it into class Flower and reuse.

File: src/com/store/flowers/FlowerStore.java (hunk 12 above)
Comment (taras309) [reply src/com/store/flowers/FlowerStore.java]: If you're talking about `flatMap` then yes, I do, coz I create three different streams for each flower then create the resulting stream using `Stream.of` and using `flatMap` I fetch each object from their stream to the resulting stream.

**Pull request merged**